src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+            end
+            all_typed_args[arg.external_name] << program.semantic(def_val).type rescue nil
+          end
+        end
+
+        # If a given collection of def_instances has a splat defined AND at least one def_instance didn't have a type for it,
+        # then we can't add types to the signature.
+        # https://crystal-lang.org/reference/1.14/syntax_and_semantics/type_restrictions.html#splat-type-restrictions
+        if @type_splats && (splat_arg = splat_arg_name) && encountered_non_splat_arg_def_instance
+          @warnings << "Not adding type restriction for splat #{splat_arg}, found empty splat call: #{parsed.location}"
+          all_typed_args.delete(splat_arg)
+        end
+        if @type_double_splats && (named_arg = named_arg_name) && encountered_non_double_splat_arg_def_instance
+          @warnings << "Not adding type restriction for double splat #{named_arg}, found empty deouble splat call: #{parsed.location}"
+          all_typed_args.delete(named_arg)
+        end
+
+        # Convert each set of types into a single ASTNode (for easier printing) representing those types
+        all_args = all_typed_args.compact_map do |name, type_set|
+          compacted_types = filter_no_return(type_set)
+
+          {name, to_ast(compacted_types)}
+        end.to_h
+
+        # Similar idea for return_type to get into an easier to print state
+        returns = filter_no_return(def_instances.compact_map do |inst|
+          resolve_type(inst)
+        end.uniq!)
+
+        return_type = to_ast(returns)
+
+        {parsed.location.to_s, Signature.new(
+          name: parsed.name,
+          return_type: return_type,
+          location: parsed.location.not_nil!,
+          args: all_args
+        )}
+      end.to_h
+    end
+
+    # Given a list of types, wrap them in a ASTNode appropriate for printing that type out
+    private def to_ast(types : Array(Crystal::Type))
+      flattened = flatten_types(types)
+      case flattened.size
+      when 1
+        # Use var to communicate a single type name
+        Crystal::Var.new(type_name(flattened[0]))
+      when 2
+        if flattened.includes?(program.nil)
+          # One type is Nil, so write this using the slightly more human readable format with a '?' suffix
+          not_nil_type = flattened.reject(&.==(program.nil))[0]
+          Crystal::Var.new("#{not_nil_type}?")
+        else
+          Crystal::Union.new(flattened.map { |t| Crystal::Var.new(type_name(t)).as(Crystal::ASTNode) })
+        end
+      else
+        Crystal::Union.new(flattened.map { |t| Crystal::Var.new(type_name(t)).as(Crystal::ASTNode) })
+      end
+    end
+
+    def flatten_types(types : Array(Crystal::Type)) : Array(Crystal::Type)
+      types.map do |type|
+        type.is_a?(Crystal::UnionType) ? flatten_types(type.concrete_types) : type
+      end.flatten.uniq
+    end
+
+    def type_name(type : Crystal::Type) : String
+      type.to_s.gsub(/:Module$/, ".class").gsub("+", "")
```

Comment:
This was an interesting discovery: metatypes of classes can be created using the `.class`, but the syntax for creating a metatype of a module (also `.class`) isn't the same string representation that the compiler uses. The more you know